neqo-udp/src/lib.rs

```diff
@@ -8,37 +8,133 @@
 #![allow(clippy::missing_panics_doc)] // Functions simply delegate to tokio and quinn-udp.
 
 use std::{
+    cell::RefCell,
     io::{self, IoSliceMut},
-    net::{SocketAddr, ToSocketAddrs},
+    net::SocketAddr,
     slice,
 };
 
-use neqo_common::{Datagram, IpTos};
+use neqo_common::{qdebug, Datagram, IpTos};
 use quinn_udp::{EcnCodepoint, RecvMeta, Transmit, UdpSocketState};
-use tokio::io::Interest;
 
 /// Socket receive buffer size.
 ///
 /// Allows reading multiple datagrams in a single [`Socket::recv`] call.
 const RECV_BUF_SIZE: usize = u16::MAX as usize;
 
-pub struct Socket {
+std::thread_local! {
+    static RECEIVE_BUFFER: RefCell<Vec<u8>> = RefCell::new(vec![0; RECV_BUF_SIZE]);
+}
+
+pub struct Socket<S> {
+    state: UdpSocketState,
     #[allow(unknown_lints)] // available with Rust v1.75
     #[allow(clippy::struct_field_names)]
-    socket: tokio::net::UdpSocket,
-    state: UdpSocketState,
-    recv_buf: Vec<u8>,
+    socket: S,
+}
+
+impl<S> Socket<S> {
+    fn send_inner(
+        state: &UdpSocketState,
+        socket: quinn_udp::UdpSockRef<'_>,
+        d: &Datagram,
+    ) -> io::Result<()> {
+        let transmit = Transmit {
+            destination: d.destination(),
+            ecn: EcnCodepoint::from_bits(Into::<u8>::into(d.tos())),
+            contents: d,
+            segment_size: None,
+            src_ip: None,
+        };
+
+        state.send(socket, &transmit)?;
+
+        qdebug!(
+            "sent {} bytes from {} to {}",
+            d.len(),
+            d.source(),
+            d.destination()
+        );
+
+        Ok(())
+    }
+
+    fn recv_inner(
+        local_address: &SocketAddr,
+        state: &UdpSocketState,
+        socket: quinn_udp::UdpSockRef<'_>,
+    ) -> Result<Vec<Datagram>, io::Error> {
+        let mut meta = RecvMeta::default();
+
+        let dgrams =
+            RECEIVE_BUFFER.with_borrow_mut(|recv_buf| -> Result<Vec<Datagram>, io::Error> {
+                state.recv(
+                    socket,
+                    &mut [IoSliceMut::new(recv_buf)],
+                    slice::from_mut(&mut meta),
+                )?;
+
+                Ok(recv_buf[0..meta.len]
+                    .chunks(meta.stride.min(recv_buf.len()))
+                    .map(|d| {
+                        qdebug!(
+                            "received {} bytes from {} to {}",
+                            d.len(),
+                            meta.addr,
+                            local_address,
+                        );
+                        Datagram::new(
+                            meta.addr,
+                            *local_address,
+                            meta.ecn.map(|n| IpTos::from(n as u8)).unwrap_or_default(),
+                            None, // TODO: get the real TTL https://github.com/quinn-rs/quinn/issues/1749
```

Comment:
I wonder if we should remove the TTL stuff from neqo until quinn actually supports it...

Reply:
Separate issue, but I'm not sure that we really need TTL, so I'd be happy to see it removed.

Reply:
Sounds good to me. I created https://github.com/mozilla/neqo/pull/1940.